exquisite_corpus/spm_language_model_utils.py

```diff
@@ -0,0 +1,144 @@
+import numpy as np
+import torch
+import torch.nn as nn
+
+from exquisite_corpus.spm_dataset import SpmIdsBatchMaker
+
+
+class SpmIdsBatchMakerForLanguageModels(SpmIdsBatchMaker):
+    """
+    Subclass of SpmIdsBatchMaker which produces batches of items from its
+    underlying dataset along with labels suitable for training a language
+    model.  For each sequence of ids in a batch, the corresponding sequence
+    of labels is the sequence of following ids which the model is learning
+    to predict (so for the sequence i0, i1, i2 the labels are i1, i2, i3,
+    which means we must return batches of sequences shorter by one id than
+    the actual sequences in the dataset).  The only difference from a plain
+    SpmIdsBatchMaker is that the collated batches returned are pairs of tensors,
+    the first being the (shortened) data items from the data set and the second
+    being the corresponding labels.
+
```

Comment:
Does this ever model the probability distribution of the first token in the text -- so, the probability of each token given no previous tokens? If not, would prepending a "start of text" token solve the problem?

Reply:
No, we weren't modeling the distribution of the first tokens.  Given that we break long inputs up during dataset creation, I think the right way to fix that is to prepend the SentencePiece special start id to each sentence during dataset creation, before splitting up long sentences (and to add the special end id then, too), and to leave the code here alone.  I've implemented that.